File: src/my_cpp_pkg/src/add_two_ints_server.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"

class AddTwoIntsServer : public rclcpp::Node
{
public:
    AddTwoIntsServer() : Node("add_two_ints_server")
    {
        // Create a service server
        service_ = this->create_service<example_interfaces::srv::AddTwoInts>(
            "add_two_ints", std::bind(&AddTwoIntsServer::add, this, std::placeholders::_1, std::placeholders::_2));
        
    }

private:
    // Add your private methods and members here
    rclcpp::Service<example_interfaces::srv::AddTwoInts>::SharedPtr service_ ;
    
    void add(const example_interfaces::srv::AddTwoInts::Request::SharedPtr request,
              example_interfaces::srv::AddTwoInts::Response::SharedPtr response)
    {
        // Perform the addition
        response->sum = request->a + request->b;
        RCLCPP_INFO(this->get_logger(), "Incoming request: a=%ld b=%ld", request->a, request->b);
        RCLCPP_INFO(this->get_logger(), "Sending back response: [%ld]", response->sum);
    }
};


int main(int argc, char * argv[])
{

    rclcpp::init(argc, argv);

    auto node = std::make_shared<AddTwoIntsServer>();
    // RCLCPP_INFO(node3->get_logger(), "Hello World! from node2");

    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;

}

File: src/my_py_pkg/my_py_pkg/add_two_ints_server.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts


#create a node with the name 'my_first_node', inside the node we will write the code to publish and subscribe to topics
# this node sits inside the 'my_py_pkg' package
#so a pakage is a collection of nodes or a single node

class AddIntsServerNode(Node):
    def __init__(self):
        super().__init__('add_two_ints_server')

        self.server = self.create_service(AddTwoInts, 'add_two_ints', self.add_two_ints_callback)


    #method names timer_callback, this method will be called every 1 second
    def add_two_ints_callback(self, request, response):
        self.get_logger().info('Incoming request: a=%d b=%d' % (request.a, request.b))
        response.sum = request.a + request.b
        self.get_logger().info('Outgoing response: sum=%d' % (response.sum))
        return response


# this is the main function that will be called when we run the node    
def main(args=None):
    rclpy.init(args=args)

    node = AddIntsServerNode()
    rclpy.spin(node) # this will block the program until the node is shutdown
    rclpy.shutdown() # this will shutdown the node


if __name__ == '__main__':
    main()